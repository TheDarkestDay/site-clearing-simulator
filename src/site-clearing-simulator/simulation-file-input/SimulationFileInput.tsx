import { Card, Typography, CardContent } from '@material-ui/core'
import React, { useRef } from 'react';
import { useDispatch } from 'react-redux';
import { startSimulation } from '../store/store-slice';
import styles from './styles.module.css';

export const SimulationFileInput = () => {
  const dispatch = useDispatch();
  const fileFieldRef = useRef<HTMLInputElement>(null);

  const handleFileUpload = async () => {
    const fileField = fileFieldRef.current;
    if (fileField === null) {
      return;
    }

    const uploadedFiles = fileField.files;
    if (uploadedFiles != null) {
      const mapFile = uploadedFiles[0];
      const textContent = await mapFile.text();

      const siteRows = textContent.split('\n');
      const siteMap = [
        ...siteRows.map((row) => row.split(''))
      ];

      dispatch(startSimulation(siteMap));
    }
  };

  return (
    <Card>
      <CardContent>
        <form>
          <Typography className={styles.fileFieldLabel} component="label" htmlFor="simulationFile" gutterBottom>
            Please, provide a file to start simulation:
          </Typography>

          <input ref={fileFieldRef} id="simulationFile" type="file" accept="text/plain" onChange={handleFileUpload}/>
        </form>
      </CardContent>
    </Card>
  );
};