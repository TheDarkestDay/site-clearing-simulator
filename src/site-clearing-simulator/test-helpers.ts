import { SiteClearingSimulatorState, BulldozerDirection } from './store/store-slice';

export const getTestState = (
  overrides: Partial<SiteClearingSimulatorState> = {}
): SiteClearingSimulatorState => {
  return {
    map: [],
    isStarted: false,
    isStopped: false,
    bulldozerDirection: BulldozerDirection.Right,
    bulldozerPosition: {
      x: 0,
      y: -1,
    },
    isPreservedTreeRemoved: false,
    fuelUsed: 0,
    eventsLog: [],
    ...overrides,
  };
};