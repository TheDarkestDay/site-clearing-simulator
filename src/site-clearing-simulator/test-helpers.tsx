import { createStore } from '@reduxjs/toolkit';
import { create } from 'react-test-renderer';
import React from 'react';
import { Provider } from 'react-redux';
import { GlobalState, rootReducer } from '../store';
import { SiteClearingSimulatorState, BulldozerDirection, initialState } from './store/store-slice';

export const getTestState = (
  overrides: Partial<SiteClearingSimulatorState> = {}
): SiteClearingSimulatorState => {
  return {
    map: [],
    isStarted: false,
    isStopped: false,
    bulldozerDirection: BulldozerDirection.Right,
    bulldozerPosition: {
      x: 0,
      y: -1,
    },
    isPreservedTreeRemoved: false,
    fuelUsed: 0,
    eventsLog: [],
    ...overrides,
  };
};

export const renderWithStore = (component: React.ReactNode, state: Partial<GlobalState['siteClearingSimulator']> = {}) => {
  const testStore = createStore(rootReducer, {
    siteClearingSimulator: {
      ...initialState,
      ...state,
    }
  });
  const consumer = (
    <Provider store={testStore}>
      {component}
    </Provider>
  );

  return create(consumer);
}