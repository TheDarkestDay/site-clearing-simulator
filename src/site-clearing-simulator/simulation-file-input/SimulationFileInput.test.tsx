import { renderWithStore } from '../test-helpers';
import { SimulationFileInput } from './SimulationFileInput';

describe('SimulationFileInput component', () => {
  it('should match snapshot', () => {
    const wrapper = renderWithStore(<SimulationFileInput />);

    expect(wrapper).toMatchSnapshot();
  });
});