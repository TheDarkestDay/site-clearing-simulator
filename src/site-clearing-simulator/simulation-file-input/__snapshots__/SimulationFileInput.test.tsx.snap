// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimulationFileInput component should match snapshot 1`] = `
<div
  className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiCardContent-root"
  >
    <form>
      <label
        className="MuiTypography-root fileFieldLabel MuiTypography-body1 MuiTypography-gutterBottom"
        htmlFor="simulationFile"
      >
        Please, provide a file to start simulation:
      </label>
      <input
        accept="text/plain"
        id="simulationFile"
        onChange={[Function]}
        type="file"
      />
    </form>
  </div>
</div>
`;
